### PR TITLE
refactor: update `build.sh` command-line arguments

### LINKED ISSUE #969: Enhance `buildsh` with args support for OS, version, and auto-continue
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.

### Description

Refactor the build.sh script to allow specifying operating system, distribution, version, and edition via command-line arguments.

Additionally, update the help menu to reflect the new argument options and implement an --auto-continue option that allows users to bypass manual prompts, streamlining the automation process.

### Use Case(s)

Ansible/AWX Integration: This enhancement allows users to integrate the build.sh script seamlessly into Ansible playbooks or AWX workflows by specifying the exact operating system, distribution, version, and edition without manual input.

### Potential Configuration

Command-line arguments: --os, --dist, --version, --edition, --auto-continue
Help Menu Update: Enhanced to include information about these new flags.

### References

_No response_

### PR DESCRIPTION
**Pull request by ClemCreator**
# 

## Summary of Pull Request

- **Refactored the `build.sh` script** to support operating system selection, distribution, version, and edition via command-line arguments.
- **Updated the help menu** to reflect the new options and parameters.
- **Implemented an `--auto-continue` option** to skip manual prompts, enabling smoother automation.
- **Added error handling** for incorrect parameter inputs.

## Type of Pull Request

- [ ] This is a bugfix. `type/bug`
- [x] This is an enhancement or feature. `type/feature`
- [x] This is a refactoring update. `type/refactor`
- [ ] This is a documentation update. `type/docs`
- [ ] This is a chore. `type/chore`
- [ ] This is something else.
      Please describe:

## Related to Existing Issues

Closes #969

## Test and Documentation Coverage

- [x] Tests have been completed.
- [x] Documentation has been added or updated.

## Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -394,8 +441,14 @@ select_os() {
     done
 }
 
+
 select_distribution() {
-    # Check if the selected guest operating system is Linux or Windows.
+    if [[ -n "$dist" ]]; then
+        echo "Distribution sélectionnée : $dist"
```
Comment (tenthirtyam) [inline build.sh]: Please use "selection".

File: build.sh [hunk 2]
```diff
@@ -454,7 +507,23 @@ select_distribution() {
 
 # This function selects the version based on the guest operating system's distribution or type.
 select_version() {
-    # Check if the selected guest operating system is Windows.
+    if [[ -n "$version" ]]; then
+        echo "Version sélectionnée : $version"
```
Comment (tenthirtyam) [inline build.sh]: Please use "selection".

File: build.sh [hunk 3]
```diff
@@ -454,7 +507,23 @@ select_distribution() {
...
+        if [[ "$dist" == *"Windows"* ]]; then
+            # Si l'édition est fournie, passer directement à select_build
```
Comment (tenthirtyam) [inline build.sh]: Please use comments in English.

File: build.sh [hunk 4]
```diff
@@ -500,6 +568,12 @@ select_version() {
 }
 
 select_edition() {
+    if [[ -n "$edition" ]]; then
+        echo "Édition sélectionnée : $edition"
```
Comment (tenthirtyam) [inline build.sh]: Please use "selection".

File: build.sh [hunk 5]
```diff
@@ -562,22 +636,31 @@ select_build() {
...
-                break
-                ;;
-            [q]*)
-                # Quit the script.
-                exit 0
-                ;;
-            esac
-        done
+
+        # Vérification si auto-continue est activé
```
Comment (tenthirtyam) [inline build.sh]: Please use comments in English.

File: build.sh [hunk 6]
```diff
@@ -562,22 +636,31 @@ select_build() {
...
+        if [ "$auto_continue" = true ]; then
+            printf "\nAuto-continue is enabled. Skipping user prompt and retrying...\n"
+            select_version  # Continue automatiquement
```
Comment (tenthirtyam) [inline build.sh]: Please use comments in English.

File: build.sh [hunk 7]
```diff
@@ -588,7 +671,15 @@ select_build() {
       printf "\n\033[33mWarning:\033[0m The branch name contains a slash ('\033[33m/\033[0m') which may cause issues with the build.\n\n"
     fi
 
+
+
     while true; do
+        if [ "$auto_continue" = true ]; then
+            # Si auto-continue est activé, on continue sans prompt
```
Comment (tenthirtyam) [inline build.sh]: Please use comments in English.

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Thanks for thr contribution!

A couple of nits:

- Please use English for stdout and comments.
- Please open and enhancement request and link this pull request to the issue per the contributing guidelines.
- Please use conventional commits per the contributing guidelines. See the commit history for examples.
- Please include any testing performed.

**Comment by tenthirtyam** [comment]
Marked as draft until CLA is completed.

File: build.sh [hunk 8]
```diff
@@ -562,22 +636,31 @@ select_build() {
...
-                break
-                ;;
-            [q]*)
-                # Quit the script.
-                exit 0
-                ;;
-            esac
-        done
+
+        # Check if auto-continue is enabled
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
        # Check if auto-continue is enabled.
```

File: build.sh [hunk 9]
```diff
@@ -562,22 +636,31 @@ select_build() {
...
+        if [ "$auto_continue" = true ]; then
+            printf "\nAuto-continue is enabled. Skipping user prompt and retrying...\n"
+            select_version  # automatically continue
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
            select_version  # Automatically continue.
```

File: build.sh [hunk 10]
```diff
@@ -588,7 +671,15 @@ select_build() {
       printf "\n\033[33mWarning:\033[0m The branch name contains a slash ('\033[33m/\033[0m') which may cause issues with the build.\n\n"
     fi
 
+
+
     while true; do
+        if [ "$auto_continue" = true ]; then
+            # If auto-continue is enabled, proceed without prompt
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
            # If auto-continue is enabled, proceed without prompt.
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]

**Comment by tenthirtyam** [comment]
Rebased from `develop` and sqashed/updated commits to:

> _refactor: update `build.sh` command-line arguments_

**Comment by ClemCreator** [comment]
- [X] > * Please use English for stdout and comments.
- [X] > * [Please open and enhancement request and link this pull request to the issue per the contributing guidelines.](https://github.com/vmware-samples/packer-examples-for-vsphere/issues/969)
- [X] > * Please useed. conventional commits per the contributing guidelines. See the commit history for examples.
- [X] > * Please include any testing performed:
	_I developed a Bash script that generates and tests all possible commands based on the `project.json` file. Additionally, I performed error testing on the parameters (see screenshots). Initially, I started with tests using Bats (Bash Automated Testing System), but thoroughly testing the entire `build.sh` script would require more time than I currently have available._


Capture d’écran 2024-10-02 151256
Capture d’écran 2024-10-02 151551

**Comment by burnsjared0415** [comment]
Please update the documentation to reflect these changes in the mark down files

File: build.sh [hunk 11]
```diff
@@ -195,6 +208,15 @@ check_dependencies() {
 }
 
 show_command=0
+os=""
+dist=""
+version=""
+edition=""
+auto_continue=false
+# auto_retry=false
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
```

File: build.sh [hunk 12]
```diff
@@ -207,6 +229,30 @@ while (("$#")); do
...
+        shift 2
+        ;;
+    --auto-continue)
+        auto_continue=true
+        shift
+        ;;
+    # --retry)
+    #     auto_retry=true
+    #     shift
+    #     ;;
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
```

File: build.sh [hunk 13]
```diff
@@ -423,6 +486,25 @@ select_distribution() {
...
+        fi
+        echo -e "\033[0;32m✔ Selected distribution:\033[0m $dist"
+        select_version
+        return
+    fi
+
+    print_title
+
+
+
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    print_title

```

File: build.sh [hunk 14]
```diff
@@ -464,6 +546,26 @@ select_version() {
...
+        else
+            select_build
+        fi
+        return
+    fi
+
+
 
     # Print the submenu.
     clear
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    fi

    # Print the submenu.
    clear
```

File: build.sh [hunk 15]
```diff
@@ -503,6 +604,22 @@ select_edition() {
...
+            done
+            exit 1
+        fi
+        echo -e "\033[0;32m✔ Selected edition:\033[0m $edition"
+        select_build
+        return
+    fi
+
+
     # Print the submenu.
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    fi

    # Print the submenu.
```

File: build.sh [hunk 16]
```diff
@@ -552,32 +669,42 @@ select_build() {
...
+                [qQ]*)
+                    exit 0
+                    ;;
+                esac
+            done
+        fi
     fi
+
+
     BUILD_PATH=${INPUT_PATH#"${script_path}/builds/"}
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    fi

    BUILD_PATH=${INPUT_PATH#"${script_path}/builds/"}
```

File: build.sh [hunk 17]
```diff
@@ -588,7 +715,15 @@ select_build() {
       printf "\n\033[33mWarning:\033[0m The branch name contains a slash ('\033[33m/\033[0m') which may cause issues with the build.\n\n"
     fi
 
+
+
     while true; do
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    fi

    while true; do
```

File: build.sh [hunk 18]
```diff
@@ -862,6 +997,8 @@ if $logging_enabled; then
     fi
 fi
 
+
+
 # Start the script with the selecting the guest operating system.
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
fi

# Start the script with the selecting the guest operating system.
```

File: build.sh [hunk 19]
```diff
@@ -74,13 +74,26 @@ info() {
...
+    printf "  \033[0;34m --help, -h, -H\033[0m       📝 Display this help message.\n\n"
+    printf "  \033[0;34m --deps, -d, -D\033[0m       ⚙ Check the dependencies.\n"
+    printf "  \033[0;34m --debug, -b, -B\033[0m      🛠 Enable debug mode.\n"
+    printf "  \033[0;34m --json, -j, -J\033[0m       📂 Specify the JSON file path.\n"
+    printf "  \033[0;34m --os\033[0m                 💻 Specify the operating system (Linux, Windows).\n"
+    printf "  \033[0;34m --dist\033[0m               📦 Specify the distribution (e.g., 'Ubuntu', 'Windows Server').\n"
+    printf "  \033[0;34m --version\033[0m            📊 Specify the version of the distribution.\n"
+    printf "  \033[0;34m --edition\033[0m            📝 Specify the edition (e.g., 'Standard', 'Datacenter').\n"
+    printf "  \033[0;34m --auto-continue\033[0m      ⏳ Automatically continue without user prompts.\n"
+    printf "  \033[0;34m --show, -s, -S\033[0m       🔍 Display the build command used to build the image.\n"
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    printf "\033[0;32m Usage\033[0m: %s [options]\n\n" "$script_name"
    printf "\033[0;34m Options:\033[0m\n"
    printf "  \033[0;34m --help, -h, -H\033[0m       Display this help message.\n\n"
    printf "  \033[0;34m --deps, -d, -D\033[0m       Check the dependencies.\n"
    printf "  \033[0;34m --debug, -b, -B\033[0m      Enable debug mode.\n"
    printf "  \033[0;34m --json, -j, -J\033[0m       Specify the JSON file path.\n"
    printf "  \033[0;34m --os\033[0m                 Specify the operating system (Linux, Windows).\n"
    printf "  \033[0;34m --dist\033[0m               Specify the distribution (e.g., 'Ubuntu', 'Windows Server').\n"
    printf "  \033[0;34m --version\033[0m            Specify the version of the distribution.\n"
    printf "  \033[0;34m --edition\033[0m            Specify the edition (e.g., 'Standard', 'Datacenter').\n"
    printf "  \033[0;34m --auto-continue\033[0m      Automatically continue without user prompts.\n"
    printf "  \033[0;34m --show, -s, -S\033[0m       Display the build command used to build the image.\n"
```

File: build.sh [hunk 20]
```diff
@@ -74,13 +74,26 @@ info() {
...
+
+    # Handling user input or exit
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
    # Handle user input or exit.
```

File: build.sh [hunk 21]
```diff
@@ -195,6 +208,15 @@ check_dependencies() {
...
+version=""
+edition=""
+auto_continue=false
+# auto_retry=false
+# Script options.
+
+
 # Script options.
 while (("$#")); do
     case "$1" in
```
Comment (tenthirtyam) [inline build.sh]:
```suggestion
auto_continue=false

# Script options.
while (("$#")); do
    case "$1" in
```

Comment (ClemCreator) [reply build.sh]: https://github.com/ClemTheStudent/packer-examples-for-vsphere/commit/aa5102d23bd5d4ac112a03545c21dfce5b2d511a

**Comment by ClemCreator** [comment]
> Please update the documentation to reflect these changes in the mark down files

https://github.com/ClemTheStudent/packer-examples-for-vsphere/commit/ad36899a424d33f2eb404ef20d50eae91de469f5

**Comment by ClemCreator** [comment]
Hi @tenthirtyam and team,

I've addressed the requested changes and updated the documentation. Let me know if anything else is needed to move this forward.

Thanks!

**Comment by tenthirtyam** [comment]
We'll need to review but also need to wait until the CLA is approved.

**Comment by tenthirtyam** [comment]
@normansden @chipchilders - Could the open CLA request be reviewed for approval?

**Review by tenthirtyam — APPROVED** [review]

**Pull request merged**